src/julia-syntax.scm

```diff
@@ -2738,15 +2752,28 @@
                               (vinfo:set-sa! vi #t))
                           (vinfo:set-asgn! vi #t)))
                e)
-             (begin (analyze-vars (caddr e) env captvars sp)
+             (begin (analyze-vars (caddr e) env captvars sp glob-assign)
                     (assert (eq? (car (cadddr e)) 'lambda))
                     (analyze-vars-lambda (cadddr e) env captvars sp
-                                         (method-expr-static-parameters e)))))
+                                         (method-expr-static-parameters e)
+                                         glob-assign))))
         ((module toplevel) e)
-        (else (for-each (lambda (x) (analyze-vars x env captvars sp))
+        (else (for-each (lambda (x) (analyze-vars x env captvars sp glob-assign))
                         (cdr e))))))
 
-(define (analyze-variables! e) (analyze-vars e '() '() '()) e)
+(define (analyze-variables! e)
+  (let ((glob-assign (table)))
+    (analyze-vars e '() '() '() glob-assign)
+    ;; if we collected any assignments to globals
+    ;; annotate them now at the toplevel
+    (let ((glob-decl (map (lambda (e) `(global ,e)) (table.keys glob-assign))))
+      (if (null? glob-decl)
+          e
+          (insert-after-meta
+            (if (and (pair? e) (eq? (car e) 'block))
+                e
+                `(block ,e))
+            glob-decl)))))
```

Comment:
Also, while having `jl_resolve_globals` do this will mostly work, it's still broken if the global code has any other side-effects, since the current scope rules can't preserve the happens-before relationship the user may have intended. Ignoring the trivial cases like code that uses `eval` or `include`, another example of where we must give the wrong behavior is if `import` is in the same block as `global`. (so `import Foo.bar; global bar() = 3` will execute in the wrong order, for example)